vendor/fontstash/fontstash.odin

```diff
@@ -0,0 +1,1234 @@
+
+	font: ^Font,
+	previousGlyphIndex: Glyph_Index,
+
+	// unicode iteration
+	utf8state: rune, // utf8
+	codepoint: rune,
+	text: string,
+	codepointCount: int,
+
+	// byte indices
+	str: int,
+	next: int,
+	end: int,
+}
+
+// push a state, copies the current one over to the next one
+PushState :: proc(using ctx: ^FontContext, loc := #caller_location) #no_bounds_check {
+	runtime.bounds_check_error_loc(loc, state_count, MAX_STATES)
+
+	if state_count > 0 {
+		states[state_count] = states[state_count - 1]
+	}
+
+	state_count += 1
+}
+
+// pop a state 
+PopState :: proc(using ctx: ^FontContext) {
+	if state_count <= 1 {
+		log.error("FONTSTASH: state underflow! to many pops were called")
+	} else {
+		state_count -= 1
+	}
+}
+
+// clear current state
+ClearState :: proc(ctx: ^FontContext) {
+	state := __getState(ctx)
+	state.size = 12
+	state.color = 255
+	state.blur = 0
+	state.spacing = 0
+	state.font = 0
+	state.ah = .LEFT
+	state.av = .BASELINE
+}
+
+__getState :: #force_inline proc(ctx: ^FontContext) -> ^State #no_bounds_check {
+	return &ctx.states[ctx.state_count - 1]
+}
+
+SetSize :: proc(ctx: ^FontContext, size: f32) {
+	__getState(ctx).size = size
+}
+
+SetColor :: proc(ctx: ^FontContext, color: [4]u8) {
+	__getState(ctx).color = color
+}
+
+SetSpacing :: proc(ctx: ^FontContext, spacing: f32) {
+	__getState(ctx).spacing = spacing
+}
+
+SetBlur :: proc(ctx: ^FontContext, blur: f32) {
+	__getState(ctx).blur = blur
+}
+
+SetFont :: proc(ctx: ^FontContext, font: int) {
+	__getState(ctx).font = font
+}
+
+SetAH :: SetAlignHorizontal
+SetAV :: SetAlignVertical
+
+SetAlignHorizontal :: proc(ctx: ^FontContext, ah: AlignHorizontal) {
+	__getState(ctx).ah = ah
+}
+
+SetAlignVertical :: proc(ctx: ^FontContext, av: AlignVertical) {
+	__getState(ctx).av = av
+}
+
+__getQuad :: proc(
+	ctx: ^FontContext,
+	font: ^Font,
+	
+	previousGlyphIndex: i32,
+	glyph: ^Glyph,
+
+	scale: f32,
+	spacing: f32,
+	
+	x, y: ^f32,
+	quad: ^Quad,
+) {
+	if previousGlyphIndex != -1 {
+		adv := f32(__getGlyphKernAdvance(font, previousGlyphIndex, glyph.index)) * scale
+		x^ += f32(int(adv + spacing + 0.5))
+	}
+
+	// fill props right
+	rx, ry, x0, y0, x1, y1, xoff, yoff, glyph_width, glyph_height: f32
+	xoff = f32(glyph.xoff + 1)
+	yoff = f32(glyph.yoff + 1)
+	x0 = f32(glyph.x0 + 1)
+	y0 = f32(glyph.y0 + 1)
+	x1 = f32(glyph.x1 - 1)
+	y1 = f32(glyph.y1 - 1)
+
+	switch ctx.location {
+		case .TOPLEFT: {
+			rx = math.floor(x^ + xoff)
+			ry = math.floor(y^ + yoff)
+			
+			quad.x0 = rx
+			quad.y0 = ry
+			quad.x1 = rx + x1 - x0
+			quad.y1 = ry + y1 - y0
+
+			quad.s0 = x0 * ctx.itw
+			quad.t0 = y0 * ctx.ith
+			quad.s1 = x1 * ctx.itw
+			quad.t1 = y1 * ctx.ith
+		}
+
+		case .BOTTOMLEFT: {
+			rx = math.floor(x^ + xoff)
+			ry = math.floor(y^ - yoff)
+
+			quad.x0 = rx
+			quad.y0 = ry
+			quad.x1 = rx + x1 - x0
+			quad.y1 = ry - y1 + y0
+
+			quad.s0 = x0 * ctx.itw
+			quad.t0 = y0 * ctx.ith
+			quad.s1 = x1 * ctx.itw
+			quad.t1 = y1 * ctx.ith
+		}
+	}
```

Comment:
Please try to adhere to the Odin styling, `case`s should not be indented and should not have an explicit `{}`.

Reply:
ah as always :smile: